### ISSUE TITLE
Activity name/notes not preserved bidirectionally in TCX files

### ISSUE THREAD

**Issue by dlenski** [cpfair/tapiriik#99]
First of all, many thanks for writing this excellent software which does a heroically good job of syncing activities between different services without loss of data.

# Issue

The TCX Activity Notes field is used by tapiriik in a non-transitive way: because of this, it's not possible to correctly preserve both the Name and Notes fields when syncing activities between services like Strava or Garmin Connect, which have both those fields, and TCX files stored using the Dropbox service.

* Extraction **from** TCX: d262d1563ef593b0e246c83f31ac7352cd270115 uses the TCX Notes field as `activity.Notes`
* Output **to** TCX: [`fef4566`](http://github.com/cpfair/tapiriik/commit/fef4566df963fcf8841480c91efac347b4b6ba5c#diff-b920479fcbe8902ecaa70760bfcab23aL243) puts `activity.Name` into the TCX Notes field. 

As a result of this, if I create a Strava Activity with **name**="Long run" and **notes**="Meandered around the hills", then sync it to Dropbox with tapiriik, I get a TCX file that looks like this:

    <Activities>
      <Activity Sport="Running">
        <Notes>Long run</Notes>

On the other hand, if I put that same TCX file in my Dropbox folder as a new activity, then sync it to Strava or Garmin Connect, I get an activity on that service with a **blank name** and **notes**="Long run".

# Proposed workaround

The TCX format is frustrating to me because it doesn't actually support a Name field for an activity, only a Notes field, and there doesn't seem to be any standard extension to differentiate the two.

My proposed workaround to allow bidirectional syncing without loss of information--admittedly a bit of a kludge--would be to use the first line of the TCX Notes field for `activity.Name` and the subsequent lines for `activity.Notes`:

    <Activities>
      <Activity Sport="Running">
        <Notes>Long run
    Meandered around the hills</Notes>

Here's a patch to do just this:

````diff
diff --git a/tapiriik/services/tcx.py b/tapiriik/services/tcx.py
index 859efc1..f1be05a 100644
--- a/tapiriik/services/tcx.py
+++ b/tapiriik/services/tcx.py
@@ -55,7 +55,10 @@ class TCXIO:
 
         xnotes = xact.find("tcx:Notes", namespaces=ns)
         if xnotes is not None:
-            act.Notes = xnotes.text
+            xnotes_lines = xnotes.splitlines()
+            act.Name = xnotes_lines[0]
+            if len(xnotes_lines)>1:
+                act.Notes = '\n'.join(xnotes.text[1:])
 
         xcreator = xact.find("tcx:Creator", namespaces=ns)
         if xcreator is not None and xcreator.attrib["{" + TCXIO.Namespaces["xsi"] + "}type"] == "Device_t":
@@ -257,8 +260,12 @@ class TCXIO:
 
         dateFormat = "%Y-%m-%dT%H:%M:%S.000Z"
 
-        if activity.Name is not None:
+        if activity.Name is not None and activity.Notes is not None:
+            etree.SubElement(act, "Notes").text = '\n'.join(activity.Name, activity.Notes)
+        elif activity.Name is not None:
             etree.SubElement(act, "Notes").text = activity.Name
+        elif activity.Notes is not None:
+            etree.SubElement(act, "Notes").text = '\n' + activity.Notes
 
         if activity.Type == ActivityType.Cycling:
             act.attrib["Sport"] = "Biking"
````

**Comment by cpfair** [comment]
Sounds good - I came across this problem back when I was first writing the TCX code, but just decided to put it off indefinitely. Can you make this into a PR?

**Comment by grzeg1** [comment]
How about using a pipe or two (||) for separation? Relying on newlines in XML does not seem so good. Using pipes (or any other pre-agreed characters) would make it more portable for platform with just names or just notes.

**Comment by dlenski** [comment]
@grzeg1, I agree that it's not a very clean solution. The right way to do it would be to define some kind of TCX extension schema that defines a new field for activity name, so that Tapiriik could include something like this in its TCX files:

````xml
        <Activity>
            <Extensions>
                <NX xmlns="http://icanhaz.skemas.com/TapiriikActivityNameExtension">
                    <Name>My awesome run</Name>
                </NX>
            </Extension>
            ...
        </Activity>
    </Activities>
</TrainingCenterDatabase>
````

**Comment by grzeg1** [comment]
I was thinking about working with what's available, just like you did. It'll be easy to enter pipes in given platform's web interface, not so much to make platform accept and preserve newline. Anyway, it's a dirty fix.

**Comment by studgeek** [comment]
Sorry if I am missing something, but why not use the file name? TCXs are already exported to DropBox with the filename starting after the first underscore?

**Comment by dlenski** [comment]
@studgeek, for one thing the filename can't encode all the characters you might have in the name.

More importantly (and this is where I ran into the problem), there's no way to **import** from TCX files in Dropbox and get the activity name and notes from there.

**Comment by studgeek** [comment]
The second part could be fixed, right? Or is there some limitation I am
missing (besides the characters)? I was just thinking using filename would
help your goal of consistency...
